src/backend/allocator/format.rs

```diff
@@ -0,0 +1,336 @@
+//! Format info tables for DRM formats.
+//!
+//! This module provides three functions, [`get_opaque`], [`has_alpha`] and [`get_bpp`].
+//!
+//! [`get_opaque`] returns the opaque alternative of a DRM format with an alpha channel.
+//!
+//! ```
+//! # use smithay::backend::allocator::Fourcc;
+//! # use smithay::backend::allocator::format::get_opaque;
+//! assert_eq!(Some(Fourcc::Xrgb8888), get_opaque(Fourcc::Argb8888));
+//! ```
+//!
+//! [`has_alpha`] returns true if the format has an alpha channel.
+//!
+//! ```
+//! # use smithay::backend::allocator::Fourcc;
+//! # use smithay::backend::allocator::format::has_alpha;
+//! assert!(has_alpha(Fourcc::Argb8888));
+//! ```
+//!
+//! [`get_bpp`] returns the number of bits per pixel of a format.
+//!
+//! ```
+//! # use smithay::backend::allocator::Fourcc;
+//! # use smithay::backend::allocator::format::get_bpp;
+//! assert_eq!(get_bpp(Fourcc::Argb8888), Some(32));
+//! ```
+
+/// Macro to generate table lookup functions for formats.
+///
+/// See the module documentation for usage details.
+macro_rules! format_tables {
+    (
+        $($fourcc: ident {
+            $(opaque: $opaque: ident,)?
+            bpp: $bpp: expr $(,)?
+        }),*
+    ) => {
+        /// Returns the opaque alternative of the specified format.
+        ///
+        /// If the format has an alpha channel, this may return the corresponding opaque format.
+        ///
+        /// Unknown formats will always return [`None`].
+        pub const fn get_opaque(
+            fourcc: $crate::backend::allocator::Fourcc,
+        ) -> Option<$crate::backend::allocator::Fourcc> {
+            match fourcc {
+                $($(
+                    $crate::backend::allocator::Fourcc::$fourcc
+                        => Some($crate::backend::allocator::Fourcc::$opaque),
+                )?)*
+                _ => None,
+            }
+        }
+
+        /// Returns true if the format has an alpha channel.
+        ///
+        /// This function may be useful to know if the alpha channel may need to be swizzled when rendering.
+        ///
+        /// Unknown formats will always return `false`.
+        pub const fn has_alpha(fourcc: $crate::backend::allocator::Fourcc) -> bool {
+            get_opaque(fourcc).is_some()
```

Comment:
Could there be alpha formats, which don't have an opaque counter part?

Reply:
Yes there is one, `Axbxgxrx106106106106`.

Although we will need to update drm-fourcc for it.

Reply:
So then this implementation doesn't work, does it?

Reply:
Yes I'll need to mark whether a format has alpha explicitly and then if it has an opaque equivalent.